### PR TITLE
[BUG FIX] Fix the visit of inner query for NestedQueryBuilder

### PR DESCRIPTION
**Pull request by zhichao-aws**
### Description
In #10110 we add visitor design pattern in QueryBuilder. For this pattern, compound query need to manage the visit for the sub-query. It gets the child visitor and use it to visit the sub-query in the `visit` method.

However, we didn't implement the `visit` method for NestedQueryBuilder. And as a result some query processors can not work for the NestedQueryBuilder. ref: https://github.com/opensearch-project/neural-search/issues/813

In this PR we implement the visit method in NestedQueryBuilder. The coding style is consistent with previous PR #10110.

### Related Issues
#10110
https://github.com/opensearch-project/neural-search/issues/813

### Check List
- [x] Functionality includes testing.
- [x] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md), if applicable.
- [x] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose), if applicable.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Comment by zhichao-aws** [comment]
Please help add a backport label for this PR

**Comment by zhichao-aws** [comment]
Hi @vibrantvarun @msfroh, could you please take a look at this PR? As you're the author and reviewer of https://github.com/opensearch-project/OpenSearch/pull/10110

**Comment by vibrantvarun** [comment]
Add the changelog

File: CHANGELOG.md [hunk 1]
```diff
@@ -67,6 +67,7 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Refactoring FilterPath.parse by using an iterative approach ([#14200](https://github.com/opensearch-project/OpenSearch/pull/14200))
 - Refactoring Grok.validatePatternBank by using an iterative approach ([#14206](https://github.com/opensearch-project/OpenSearch/pull/14206))
 - Update help output for _cat ([#14722](https://github.com/opensearch-project/OpenSearch/pull/14722))
+- Fix the visit of inner query for NestedQueryBuilder ([#14739](https://github.com/opensearch-project/OpenSearch/pull/14739))
```
Comment (vibrantvarun) [inline CHANGELOG.md:70]: Remove this change from here. and add it in below unreleased 2.x section.

Comment (zhichao-aws) [reply CHANGELOG.md:70]: It is in the `## [Unreleased 2.x] ` now. The below `[Unreleased 2.x]` a link to a GitHub comparison page

**Comment by zhichao-aws** [comment]
From the error log of `Changelog Verifier` action, we need a backport label to make it pass
> error: Please make sure that the PR has a backport label associated with it when making an entry to the CHANGELOG.md file

**Comment by gaobinlong** [comment]
Duplicated with https://github.com/opensearch-project/OpenSearch/pull/13961.

**Comment by zhichao-aws** [comment]
Steps to verify the fix:
1. start an OpenSearch cluster with this fix and neural-search plugin installed
2. register model, create index and ingest documents
```python
from opensearchpy import OpenSearch
client = OpenSearch(hosts=["localhost:9200"])

client.transport.perform_request("PUT","/_cluster/settings", body={
     "persistent": {
         "plugins.ml_commons.only_run_on_ml_node": False,
         "plugins.ml_commons.native_memory_threshold": 99,
         "plugins.ml_commons.memory_feature_enabled": True,
         "plugins.ml_commons.agent_framework_enabled": True,
         "plugins.ml_commons.allow_registering_model_via_url": True
     }
 })


task_id = client.transport.perform_request("POST","/_plugins/_ml/models/_register",params={"deploy":"true"},body={
  "name": "traced_small_model",
  "version": "1.0.0",
  "model_format": "TORCH_SCRIPT",
  "model_task_type": "text_embedding",
  "model_content_hash_value": "e13b74006290a9d0f58c1376f9629d4ebc05a0f9385f40db837452b167ae9021",
  "model_config": {
    "model_type": "bert",
    "embedding_dimension": 768,
    "framework_type": "sentence_transformers",
    "all_config": "{\"architectures\":[\"BertModel\"],\"max_position_embeddings\":512,\"model_type\":\"bert\",\"num_attention_heads\":12,\"num_hidden_layers\":6}"
  },
  "url": "https://github.com/opensearch-project/ml-commons/blob/2.x/ml-algorithms/src/test/resources/org/opensearch/ml/engine/algorithms/text_embedding/traced_small_model.zip?raw=true"
})["task_id"]

import time
# wait 10 seconds for model register and deployment
time.sleep(10)
model_id=client.transport.perform_request("GET",f"/_plugins/_ml/tasks/{task_id}")["model_id"]

# create index, ingest documents
client.transport.perform_request("PUT","/_ingest/pipeline/test-embedding-model",body={
    "description": "text embedding pipeline",
    "processors":[{
            "text_embedding":{
                "model_id":model_id,
                "field_map":{
                    "text":"embedding"
                }
            }
        }    
    ]
})

try:
    client.indices.delete(index="test")
except:
    pass

client.indices.create(index="test",body={
  "mappings": {
    "properties": {
      "text": {
        "type": "text"
      },
      "embedding": {
        "type":"nested",
        "properties":{
            "knn":{
                "type": "knn_vector",
                "dimension": 768,
                "method": {
                  "name": "hnsw",
                  "space_type": "l2",
                  "engine": "lucene",
                  "parameters": {
                    "ef_construction": 128,
                    "m": 24
                  }
                }
            }
        }
        
      }
    }
  },
  "settings": {
    "index": {
      "knn.space_type": "cosinesimil",
      "default_pipeline": "test-embedding-model",
      "knn": "true"
    }
  }
})

client.index(index="test",body={"text":["hello world"]})
client.index(index="test",body={"text":["a, b","hello"]})
```
3. verify the fix
```python
client.transport.perform_request("PUT","/_search/pipeline/test_pipeline",body=
{
  "request_processors": [
    {
      "neural_query_enricher" : {
        "default_model_id": model_id
        }
      }
  ]
})

client.transport.perform_request("GET","/test/_search?search_pipeline=test_pipeline",body=
{
  "query": {
    "nested": {
      "score_mode": "max",
      "path": "embedding",
      "query": {
        "neural": {
          "embedding.knn": {
            "query_text": "document"
          }
        }
      }
    }
  }
})
```

image

**Review by vibrantvarun — APPROVED** [review]
LGTM🚀

**Review by kotwanikunal — APPROVED** [review]

**Review by ryanbogan — APPROVED** [review]

File: server/src/main/java/org/opensearch/index/query/NestedQueryBuilder.java [hunk 2]
```diff
@@ -505,4 +506,12 @@ public TopDocsAndMaxScore topDocs(SearchHit hit) throws IOException {
             }
         }
     }
+
+    @Override
+    public void visit(QueryBuilderVisitor visitor) {
+        visitor.accept(this);
+        if (query != null) {
```
Comment (martin-gaievski) [inline server/src/main/java/org/opensearch/index/query/NestedQueryBuilder.java:513]: Why is the check for query not being null not present in previous PRs for other query clause builders? Is this a new scenario, or is the inner nested query different from other queries? If the latter is true, please add a comment to the code specifying why we need this check for nested

Comment (vibrantvarun) [reply server/src/main/java/org/opensearch/index/query/NestedQueryBuilder.java:513]: We have kept it there for 2 reasons
1. To pass unit tests
2. The same is been followed in other query builders https://github.com/opensearch-project/OpenSearch/pull/10110/files#diff-056910f9008c16e48b353a045b19d604464d8dbf03992cda29e128af92afec8c

Comment (martin-gaievski) [reply server/src/main/java/org/opensearch/index/query/NestedQueryBuilder.java:513]: reg 1 - what exact test do you mean, one that has been added in the PR or something from existing codebase?
reg 2 - for span query those `exclude` and `include` sub-queries are special case, if something like that is for nested better have a comment, this is small PR with only one query, easier to do comparing to initial PR where you've  touched multiple queries

**Comment by zhichao-aws** [comment]
Hi @kotwanikunal , could you please help merge this PR?

**Comment by kotwanikunal** [comment]
@zhichao-aws Can you please rebase? It looks like you have conflicts.

**Comment by zhichao-aws** [comment]
> @zhichao-aws Can you please rebase? It looks like you have conflicts.

Sure

**Comment by vibrantvarun** [comment]
@zhichao-aws what is the status of this PR?

**Review by dblock — APPROVED** [review]

**Comment by dblock** [comment]
I had marked this for 2.16 as it's a pretty straightforward bug fix that we should have merged a while ago.

**Comment by dblock** [comment]
@zhichao-aws can you please backport this manually to 2.x? Thanks.

**Comment by zhichao-aws** [comment]
Hi @dblock , I created PR to backport 2.x #14967 and backport 2.16 #14968 manually. If the backport 2.16 one is unneccessary feel free to close it. Thanks!

**Pull request merged**